src/main/java/buildcraft/factory/BlockTank.java

```diff
@@ -88,123 +88,232 @@ public IIcon getIconAbsolute(IBlockAccess iblockaccess, int i, int j, int k, int
-                    entityplayer.addChatComponentMessage(
-                            new ChatComponentText(
-                                    "Amount: " + tank.getTankInfo(ForgeDirection.UNKNOWN)[0].fluid.amount + " mB"));
+        }
+
+        // Pull from tank -) cell (stack friendly)
+        if (sneaking) {
+            FluidStack inCell = cont.getFluid(work);
+            FluidStack stackFluid = bottom.getTankInfo(ForgeDirection.UNKNOWN)[0].fluid;
+            net.minecraftforge.fluids.Fluid target = (inCell != null && inCell.amount > 0) ? inCell.getFluid()
+                    : (stackFluid != null ? stackFluid.getFluid() : null);
+
+            if (target != null) {
+                int accept = cont.fill(work, new FluidStack(target, INF_MB), false);
+                if (accept > 0) {
+                    // tank stack protection consistency
+                    FluidStack drained = bottom.drain(ForgeDirection.UNKNOWN, new FluidStack(target, accept), true);
+                    if (drained != null && drained.amount > 0) {
+                        if (creative) {
+                            moved = drained.amount; // tank changed, item remains
+                        } else {
+                            int filled = cont.fill(work, drained, true);
+                            moved = filled;
+
+                            // remainder logic
+                            if (filled < drained.amount) {
+                                FluidStack rem = drained.copy();
+                                rem.amount = drained.amount - filled;
+                                bottom.fill(ForgeDirection.UNKNOWN, rem, true);
+                            }
+                        }
+                    }
                 }
             }
         }
 
-        return false;
+        if (moved > 0) {
+            final int slot = player.inventory.currentItem;
+
+            if (creative) {
+                player.inventory.setInventorySlotContents(slot, held);
+                syncPlayerAndTanks(world, x, y, z, player, clicked, bottom);
+                return true;
+            }
+
+            if (held.stackSize > 1) {
+                held.stackSize -= 1;
+                player.inventory.setInventorySlotContents(slot, held);
+                addOrDrop(player, work);
+            } else {
+                player.inventory.setInventorySlotContents(slot, work);
+            }
+
+            syncPlayerAndTanks(world, x, y, z, player, clicked, bottom);
+            return true;
+        }
+        return true;
+    }
+
+    private static TileTank getBottomTank(World world, int x, int y, int z) {
+        int boty = y;
+        while (true) {
+            TileEntity below = world.getTileEntity(x, boty - 1, z);
+            if (below instanceof TileTank) boty--;
+            else break;
+        }
+        TileEntity te = world.getTileEntity(x, boty, z);
+        return (te instanceof TileTank) ? (TileTank) te : null;
```

Comment:
you can easily avoid the last `getTileEntity` call. Don't know if it's in any way expensive though, but looks a bit scary for me
```java
int boty = y;
TileTank bottomTank;

while (true) {
    TileEntity below = world.getTileEntity(x, --boty, z);
    if (below instanceof TileTank) {
        bottomTank = below;
    } else {
        break;
    }
}

return bottomTank;
```